### PR TITLE
refactor: convert Go to new constraint setup

### PR DESCRIPTION
**Pull request by jonaslagoni**
**Description**
This PR converts the Go generator to the new constraint setup introduced in #725

### CODE REVIEW

**Review by Samridhi-98 — CHANGES_REQUESTED** [review]
Kindly remove the code smell 😅

File: src/generators/go/constrainer/EnumConstrainer.ts [hunk 1]
```diff
@@ -22,44 +23,27 @@ export const DefaultEnumKeyConstraints: ModelEnumKeyConstraints = {
...
 };
 
-export type EnumKeyContext = {
-  enumKey: string,
-  constrainedEnumModel: ConstrainedEnumModel,
-  enumModel: EnumModel
-}
-export type EnumConstraintType = (context: EnumKeyContext, constraints?: ModelEnumKeyConstraints) => string;
+export function defaultEnumKeyConstraints(customConstraints?: Partial<ModelEnumKeyConstraints>): EnumKeyConstraint {
+  const constraints = {...DefaultEnumKeyConstraints, ...customConstraints};
```
Comment (magicmatatjahu) [inline src/generators/go/constrainer/EnumConstrainer.ts:31]:
```suggestion
export function defaultEnumKeyConstraints(customConstraints: Partial<ModelEnumKeyConstraints> = {}): EnumKeyConstraint {
  const constraints = {...DefaultEnumKeyConstraints, ...customConstraints};
```

You cannot destroy the `undefined/null`.

File: src/generators/go/constrainer/ModelNameConstrainer.ts [hunk 2]
```diff
@@ -22,41 +23,20 @@ export const DefaultModelNameConstraints: ModelNameConstraints = {
...
-    }
-    if (customConstraints.NAMING_FORMATTER !== undefined) {
-      constraints.NAMING_FORMATTER = customConstraints.NAMING_FORMATTER;
-    }
-    if (customConstraints.NO_RESERVED_KEYWORDS !== undefined) {
-      constraints.NAMING_FORMATTER = customConstraints.NO_RESERVED_KEYWORDS;
-    }
-  }
+export function defaultModelNameConstraints(customConstraints?: Partial<ModelNameConstraints>): ModelNameConstraint {
+  const constraints = {...DefaultModelNameConstraints, ...customConstraints};
```
Comment (magicmatatjahu) [inline src/generators/go/constrainer/ModelNameConstrainer.ts:31]:
```suggestion
export function defaultModelNameConstraints(customConstraints: Partial<ModelNameConstraints> = {}): ModelNameConstraint {
  const constraints = {...DefaultModelNameConstraints, ...customConstraints};
```

File: src/generators/go/constrainer/PropertyKeyConstrainer.ts [hunk 3]
```diff
@@ -21,43 +23,23 @@ export const DefaultPropertyKeyConstraints: PropertyKeyConstraintOptions = {
...
 
-export type PropertyKeyContext = {
-  propertyKey: string,
-  constrainedObjectModel: ConstrainedObjectModel,
-  objectModel: ObjectModel
-}
-
-export type PropertyKeyConstraintType = (context: PropertyKeyContext) => string;
+export function defaultPropertyKeyConstraints(customConstraints?: Partial<PropertyKeyConstraintOptions>): PropertyKeyConstraint {
+  const constraints = {...DefaultPropertyKeyConstraints, ...customConstraints};
```
Comment (magicmatatjahu) [inline src/generators/go/constrainer/PropertyKeyConstrainer.ts:32]:
```suggestion
export function defaultPropertyKeyConstraints(customConstraints: Partial<PropertyKeyConstraintOptions> = {}): PropertyKeyConstraint {
  const constraints = {...DefaultPropertyKeyConstraints, ...customConstraints};
```

**Review by magicmatatjahu — CHANGES_REQUESTED** [review]

File: src/generators/go/constrainer/EnumConstrainer.ts (hunk 1 above)
Comment (jonaslagoni) [reply src/generators/go/constrainer/EnumConstrainer.ts:31]: The current setup works fine 🙂 

`defaultEnumKeyConstraints(undefined);` works as expected. 

I have added the following test:

```
    test('should be able to handle undefined', () => {
      const constrainFunction = defaultEnumKeyConstraints(undefined);
      const value = constrainFunction({enumModel, constrainedEnumModel, enumKey: 'TEST'});
      expect(value).toEqual('TEST');
    });
```

Comment (magicmatatjahu) [reply src/generators/go/constrainer/EnumConstrainer.ts:30]: 🤣  I always think that undefined/null will throw error

**Comment by jonaslagoni** [comment]
> Kindly remove the code smell

@Samridhi-98 done 😆 

Think Sonar choked a bit before as those issues should not be present. Thanks for pointing them out 👍

**Review by Samridhi-98 — APPROVED** [review]
LGTM 🚀

**Review by magicmatatjahu — APPROVED** [review]

**Comment by jonaslagoni** [comment]
Thanks @Samridhi-98 @magicmatatjahu 👍

**Pull request merged**